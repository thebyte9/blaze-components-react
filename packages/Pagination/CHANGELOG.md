# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.8.0-alpha.73](https://github.com/thebyte9/blaze-components-react/compare/v0.8.0-alpha.72...v0.8.0-alpha.73) (2023-12-20)


### Bug Fixes

* publish package ([1780d78](https://github.com/thebyte9/blaze-components-react/commit/1780d781a4d81c457751ff76a0be34448e9568cc))





# [0.8.0-alpha.72](https://github.com/thebyte9/blaze-components-react/compare/v0.8.0-alpha.71...v0.8.0-alpha.72) (2023-12-14)

**Note:** Version bump only for package @blaze-react/pagination





# [0.8.0-alpha.71](https://github.com/thebyte9/blaze-components-react/compare/v0.8.0-alpha.70...v0.8.0-alpha.71) (2023-12-14)


### Bug Fixes

* render simple pagination if no count ([#682](https://github.com/thebyte9/blaze-components-react/issues/682)) ([896445a](https://github.com/thebyte9/blaze-components-react/commit/896445abfc0b030783716f08b4b49d41e09efc9a))





# [0.8.0-alpha.70](https://github.com/thebyte9/blaze-components-react/compare/v0.8.0-alpha.69...v0.8.0-alpha.70) (2023-12-06)


### Features

* add pagination component ([#628](https://github.com/thebyte9/blaze-components-react/issues/628)) ([baffd54](https://github.com/thebyte9/blaze-components-react/commit/baffd542b4ed6819eb93343e6b13951af5adcc68))
