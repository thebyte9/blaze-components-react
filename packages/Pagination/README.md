## Description

The Pagination component enables the user to select a specific page from a range of pages.

## Usage

- Required

```js
<Input
  label="Text input required"
  placeholder="Placeholder text"
  onChange={({ event, value }) => {}}
  modifier="full-width"
  validationMessage="Email address is required"
  error
  required
/>
```

- Disabled

```js
<Input
  label="Text input disabled"
  placeholder="Placeholder text"
  onChange={({ event, value }) => {}}
  type="password"
  hideTypeToggle
  disabled
/>
```

- Password toggle type

```js
<Input
  label="Password - show hide"
  placeholder="******"
  onChange={({ event, value }) => {}}
  type="password"
  value="Lorem ipsum"
/>
```

## API

##### Pagination can receive a number of `props` as follow:

| NAME              |   TYPE   |        DEFAULT         |
| :---------------- | :------: | :--------------------: |
| label             |  String  |         empty          |
| type              |  String  |          text          |
| value             |  String  |         empty          |
| validationMessage |  String  | This field is required |
| disabled          | Boolean  |         false          |
| required          | Boolean  |         false          |
| hideTypeToggle    | Boolean  |         false          |
| error             | Boolean  |         false          |
| onChange          | Function |        () => {}        |
