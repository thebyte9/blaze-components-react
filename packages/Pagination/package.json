{
  "name": "@blaze-react/pagination",
  "version": "0.8.0-alpha.60",
  "description": "",
  "main": "./dist/main.js",
  "scripts": {
    "build": "NODE_ENV=production webpack --config webpack.config.js",
    "build-storybook": "build-storybook -c .storybook -o docs-build",
    "lint": "eslint '*/**/*.{ts,tsx}' --ignore-path ../../.eslintignore --fix",
    "test": "jest --passWithNoTests",
    "test:coverage": "jest --coverage --passWithNoTests",
    "pre-publish": "yarn build && yarn lint && yarn test",
    "bundle-analyser": "node ../../node_modules/webpack-bundle-analyzer/lib/bin/analyzer.js ./stats/stats.json"
  },
  "keywords": [],
  "author": "byte9",
  "license": "GPL-3.0",
  "devDependencies": {
    "react": "^18.0.2",
    "react-dom": "^18.0.2",
    "webpack": "^5.28.0",
    "webpack-cli": "^4.6.0"
  },
  "peerDependencies": {
    "react": "^18.0.2",
    "react-dom": "^18.0.2"
  },
  "gitHead": "6e7fd8d59476e6d7c5ef48cb9b7bfc47c387f332"
}
