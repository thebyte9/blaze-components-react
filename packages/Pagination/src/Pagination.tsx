import React, { FunctionComponent, useEffect, useState } from 'react';

interface IPaginationProps {
  totalPages: number;
  activePage: number;
  paginationPagesPerSide: number;
  defaultRowsPerPage: number;
  handleOnPageChange: ({ rowsPerPage, currentPage }: { rowsPerPage: number, currentPage: number }) => void;
}

const Pagination: FunctionComponent<IPaginationProps> = ({
  totalPages,
  handleOnPageChange,
  activePage,
  paginationPagesPerSide,
  defaultRowsPerPage
}): JSX.Element => {
  const [currentPage, setCurrentPage] = useState(activePage)
  const [rowsPerPage, setRowsPerPage] = useState(defaultRowsPerPage)

  useEffect(() => {
    handleOnPageChange({ currentPage, rowsPerPage })
  }, [currentPage, handleOnPageChange, rowsPerPage])

  const getPaginationNumbers = () => {
    const { nextPages, prevPages } = getPagesPerSide();
    const numbers: number[] = [];
    for (let number = prevPages; number <= nextPages; number++) {
      numbers.push(number);
    }
    return numbers;
  };

  const getPagesPerSide = () => {
    let prevPages: number = currentPage - paginationPagesPerSide;
    let nextPages: number = currentPage + paginationPagesPerSide;
    if (prevPages <= 0) prevPages = 1;
    if (nextPages > totalPages) nextPages = totalPages;
    return { nextPages, prevPages };
  };

  const handleRowsPerPage = ({ target: { value } }) => {
    const totalRows = parseInt(value) || 0;
    setRowsPerPage(totalRows);
  }

  const nextPage = () => currentPage < totalPages && setCurrentPage(currentPage + 1);

  const prevPage = () => currentPage > 1 && setCurrentPage(currentPage - 1);

  const getItemClassName = (number: number) =>
    `pagination__item ${number === currentPage ? "pagination__item--active" : ""
    }`;

  const currentNumbers: number[] = getPaginationNumbers();

  return (
    <div className="pagination">
      <div>
        <span>Displaying</span>
        <input
          min="0"
          className="pagination__input"
          type="number"
          placeholder="Go to"
          value={rowsPerPage}
          onChange={handleRowsPerPage}
        />
        <span>rows per page</span>
      </div>
      <ul className="pagination">
        <li className="pagination__item pagination__item--icon" onClick={prevPage}>
          &lsaquo;
        </li>
        {currentNumbers.map((number) => (
          <li
            key={`${number}`}
            className={getItemClassName(number)}
            onClick={() => {
              setCurrentPage(number);
            }}
          >
            {number}
          </li>
        ))}
        <li>...</li>
        <li className="pagination__item" onClick={() => {
          setCurrentPage(totalPages);
        }}>
          {totalPages}
        </li>
        <li className="pagination__item pagination__item--icon" onClick={nextPage}>
          &rsaquo;
        </li>
      </ul>
    </div>
  );
};

Pagination.defaultProps = {
  activePage: 1,
  paginationPagesPerSide: 5,
  defaultRowsPerPage: 10
};

export default Pagination;
