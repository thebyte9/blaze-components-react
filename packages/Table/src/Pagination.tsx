import React, { Fragment, FunctionComponent, useState } from 'react';

interface IPaginationProps {
  totalPages: Number;
  currentPage: Number | any;
  paginationPagesPerSide: Number | any;
  handleOnPageChange: (...args: any[]) => void;
  query?: string;
  pageParameterName?: string;
}

const Pagination: FunctionComponent<IPaginationProps> = ({ 
  totalPages,
  handleOnPageChange,
  currentPage,
  paginationPagesPerSide,
}): JSX.Element => {
  const getPaginationNumbers = () => {
    const { nextPages, prevPages } = getPagesPerSide();
    let numbers: Number[] = [];
    for (let number: any = prevPages; number <= nextPages; number++) {
      numbers.push(number);
    }
    return numbers;
  };

  const getPagesPerSide = () => {
    let prevPages: Number = currentPage - paginationPagesPerSide;
    let nextPages: Number = currentPage + paginationPagesPerSide;
    if (prevPages <= 0) prevPages = 1;
    if (nextPages > totalPages) nextPages = totalPages;
    return { nextPages, prevPages };
  };

  const handleGoToPage = ({ target: { value } }) =>
    !!parseInt(value) && handleOnPageChange(parseInt(value));
  const nextPage = () =>
    currentPage < totalPages && handleOnPageChange(currentPage + 1);
  const prevPage = () => currentPage > 1 && handleOnPageChange(currentPage - 1);
  const getItemClassName = (number) =>
    `pagination__item ${
      number === currentPage ? "pagination__item--active" : ""
    }`;

  const currentNumbers: Number[] = getPaginationNumbers();

  return (
    <div className="pagination">
      <div>
        <span>Displaying</span>
        <input
          min="1"
          className="pagination__input"
          type="number"
          placeholder="Go to"
          value="4"
          onChange={handleGoToPage}
        />
        <span> rows per page</span>
      </div>
      <ul className="pagination">
      <li className="pagination__item pagination__item--icon" onClick={prevPage}>
        &lsaquo;
      </li>
      {currentNumbers.map((number) => (
        <li
          key={`${number}`}
          className={getItemClassName(number)}
          onClick={() => {
            handleOnPageChange(number);
          }}
        >
          {number}
        </li>
      ))}
      <li>...</li>
      <li className="pagination__item" onClick={() => {
            handleOnPageChange(totalPages);
          }}>
        {totalPages}
      </li>
      <li className="pagination__item pagination__item--icon" onClick={nextPage}>
        &rsaquo;
      </li>
      </ul>
    </div>
  );
};

Pagination.defaultProps = {
  currentPage: 1,
  paginationPagesPerSide: 5,
  pageParameterName: "page"
};

export default Pagination;
