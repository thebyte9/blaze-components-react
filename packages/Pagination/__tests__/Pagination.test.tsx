import '@testing-library/jest-dom';

import { render, screen } from '@testing-library/react';

import Pagination from '../src';
import React from 'react';
import userEvent from '@testing-library/user-event';

const defaultProps = (override = {}) => ({
    handleOnPageChange: jest.fn(),
    activePage: 1,
    paginationPagesPerSide: 5,
    defaultRowsPerPage: 10,
    totalPages: 100,
    ...override,
});

describe('Pagination component', () => {

    it("should be defined", () => {
        expect(Pagination).toBeDefined();
    });

    it('should be defined and renders correctly (snapshot)', () => {
        const { container } = render(<Pagination {...defaultProps()} />);
        expect(container).toMatchSnapshot();
    });

    it('handles page change', () => {
        const mockOnChange = jest.fn();
        render(<Pagination {...defaultProps({ handleOnPageChange: mockOnChange })} />);
        userEvent.click(screen.getByText('6'));
        expect(mockOnChange).toHaveBeenCalled();
    });

});
