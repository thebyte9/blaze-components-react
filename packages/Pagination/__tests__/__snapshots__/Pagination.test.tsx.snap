// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component should be defined and renders correctly (snapshot) 1`] = `
<div>
  <div
    class="pagination"
  >
    <div>
      <span>
        Displaying
      </span>
      <input
        class="pagination__input"
        min="0"
        placeholder="Go to"
        type="number"
        value="10"
      />
      <span>
        rows per page
      </span>
    </div>
    <ul
      class="pagination"
    >
      <li
        class="pagination__item pagination__item--icon"
      >
        ‹
      </li>
      <li
        class="pagination__item pagination__item--active"
      >
        1
      </li>
      <li
        class="pagination__item "
      >
        2
      </li>
      <li
        class="pagination__item "
      >
        3
      </li>
      <li
        class="pagination__item "
      >
        4
      </li>
      <li
        class="pagination__item "
      >
        5
      </li>
      <li
        class="pagination__item "
      >
        6
      </li>
      <li>
        ...
      </li>
      <li
        class="pagination__item"
      >
        100
      </li>
      <li
        class="pagination__item pagination__item--icon"
      >
        ›
      </li>
    </ul>
  </div>
</div>
`;

exports[`should be defined and renders correctly (snapshot) 1`] = `
<div>
  <div
    class="pagination"
  >
    <div>
      <span>
        Displaying
      </span>
      <input
        class="pagination__input"
        min="0"
        placeholder="Go to"
        type="number"
        value="10"
      />
      <span>
        rows per page
      </span>
    </div>
    <ul
      class="pagination"
    >
      <li
        class="pagination__item pagination__item--icon"
      >
        ‹
      </li>
      <li
        class="pagination__item pagination__item--active"
      >
        1
      </li>
      <li
        class="pagination__item "
      >
        2
      </li>
      <li
        class="pagination__item "
      >
        3
      </li>
      <li
        class="pagination__item "
      >
        4
      </li>
      <li
        class="pagination__item "
      >
        5
      </li>
      <li
        class="pagination__item "
      >
        6
      </li>
      <li>
        ...
      </li>
      <li
        class="pagination__item"
      >
        100
      </li>
      <li
        class="pagination__item pagination__item--icon"
      >
        ›
      </li>
    </ul>
  </div>
</div>
`;
